log4j-core/src/main/java/org/apache/logging/log4j/core/jmx/StatusLoggerAdmin.java

```diff
@@ -92,6 +80,50 @@ private static MBeanNotificationInfo createNotificationInfo() {
         return new MBeanNotificationInfo(notifTypes, name, description);
     }
 
+    /// BEGIN: Conditional `StatusListener` registration ///////////////////////////////////////////////////////////////
+
+    // `StatusLogger` contains a _fallback listener_ that defaults to a `StatusConsoleListener`.
+    // It is used to propagate logs when no listeners are available.
+    // If JMX registers itself always, unconditionally, this would render the fallback (console) listener ineffective.
+    // That is, no status logs would be written to console when `StatusLoggerAdmin` is in the classpath.
+    // To avoid this undesired behaviour, we register JMX status listener only when there is a party interested in these
+    // notifications.
+
+    @Override
+    public void addNotificationListener(
+            final NotificationListener listener, final NotificationFilter filter, final Object handback) {
+        super.addNotificationListener(listener, filter, handback);
+        registeredNotificationListeners.add(listener);
+        updateStatusListenerRegistration();
+    }
+
+    @Override
+    public void removeNotificationListener(final NotificationListener listener) throws ListenerNotFoundException {
+        super.removeNotificationListener(listener);
+        registeredNotificationListeners.remove(listener);
+        updateStatusListenerRegistration();
```

Comment:
Looking at the `NotificationBroadcasterSupport` class, the logic to add and remove listeners is more complex.

If I understand correctly:

 * NBS remembers a list of triplets `(listener, filter, handback`,
 * the `removeNotificationListener(listener)` method removes all the triplets with the give `listener` value,
 * the `removeNotificationListener(listener, filter, handback)` method remove **only** the triplet with the given values of `listener`, `filter` and `handback`.

I am not convinced we need to implement this logic, but we should either:
 * implement the complete logic,
 * or limit ourselves to registering the status listener in `addNotificationListener` and removing the status listener only when `MBeanRegistration#postDeregister` is called. This way we ignore all the intricacies of notifications.

Reply:
@ppkarwasz, thanks for the due diligence during review. I implemented the `postDeregister()` approach in 88eb8fcb415bbe7abf544467d632cde12675d375. Would you mind checking it out, please?